File: Python/103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        res = []
        if root == None:
            return res
        queue, isReverse = [root], False
        while queue != []:
            row, size = [], len(queue)
            while size != 0:
                node = queue.pop(0)
                row.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                size -= 1
            if isReverse:
                row.reverse()
            res.append(row)
            isReverse = not isReverse
        return res

File: Go/103-binary-tree-zigzag-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func reverse(nums []int) []int {
    start, end := 0, len(nums) - 1
    for start < end {
        nums[start], nums[end] = nums[end], nums[start]
        start++
        end--
    }
    return nums
}

func zigzagLevelOrder(root *TreeNode) [][]int {
    var res [][]int
    if root == nil {
        return res
    }
    queue, falg := []*TreeNode{root}, false
    for len(queue) != 0 {
        size := len(queue)
        var row []int
        for size != 0 {
            node := queue[0]
            queue, row = queue[1:], append(row, node.Val)
            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
            size--
        }
        if falg {
            row = reverse(row)
        }
        falg = !falg
        res = append(res, row)
    }
    return res
}